exploration/0477-quoted-literals.md

```diff
@@ -0,0 +1,192 @@
+- **[r2; high priority]** Minimize the need to escape characters when embedding messages in code or containers. In particular, choose a delimiter that isn't frequently used as a string delimiter in programming languages and container formats. However, note that many programming languages also provide alternative ways of delimiting strings, e.g. _raw strings_ or triple-quoted literals.
+- **[r3; medium priority]** Minimize the need to change the message in other ways than to escape some of its characters (e.g. rephrase content, use typographic apostrophes, or switch to using a second set of delimtiers).
+- **[r4; medium priority]** Don't surprise users with syntax that's too exotic. We expect quoted literals to be rare, which means fewer opportunities to get used to their syntax and remember it.
+- **[r5; low priority]** Be able to pair the opening and the closing delimiter, to aid parsers recover from syntax errors, and to leverage IDE's ability to highlight matching pairs of delimiters, to visually indicate to the user editing a message the bounds of the literal under caret. However, quoted literals are usually short and already enclosed in a placeholder (which has its own delimiters) or are outside patterns (when used as variant keys).
+
+## Constraints
+
+_What prior decisions and existing conditions limit the possible design?_
+
+- **[c1]** MessageFormat uses the backslash, `\`, as the escape sequence introducer.
+- **[c2]** Straight quotation marks, `'` and `"`, are common in content across many languages, even if other Unicode codepoints should be used in well-formatted text.
+- **[c3]** Straight quotation marks, `'` and `"`, are common as string delimiters in many programming languages.
+
+## Proposed Design
+
+_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+
+Use the vertical line character, `|`, to delimit quoted strings. For example:
+
+> ```
+> {The Unix epoch is defined as {|Thu, 01 Jan 1970 00:00:00 GMT| :datetime}.}
+> ```
+
+```abnf
+literal       = quoted / unquoted
+quoted        = "|" *(quoted-char / quoted-escape) "|"
+quoted-char   = %x0-5B         ; omit \
+              / %x5D-7B        ; omit |
+              / %x7D-D7FF      ; omit surrogates
+              / %xE000-10FFFF
+quoted-escape = backslash ( backslash / "|" )
+```
+
+By being both uncommon in text content and uncommon as a string delimiter in other programming languages, the vertical line sidesteps the "inwards" and "outwards" problems of escaping.
+
+- [r1 GOOD] Writing `"` and `'` in literals doesn't require escaping them via `\`. This means no extra `\` that need escaping.
```

Comment:
I like the scoring thing, but when writing documents like this, I tend to prefer a table for "all up" comparison somewhere in the document, perhaps at the end, e.g.

| Option | Writing | Embedding | Modification | Commonality |
|---|---|---|---|---|
| '|' | + | + | + | 0 | - |
| `'` `"` | - | - | - | + | 0 |

Reply:
I added a comparison matrix in 9d53b4d6ad4c1932d7e3e142065391844606e294.